Improve build cacheability

sandra-thieme opened the pull request:
## Description
This PR addresses a number of issues in some gradle tasks:
- ignores empty directories for task inputs
- configures task dependencies
- replace relative paths with absolute paths where applicable
- enable build caching for tasks that did not support it but should

## Motivation and Context
Some gradle tasks were configured in a way that prevented incremental builds and build caching from being used effectively. This PR addresses some of these issues. 

## How Has This Been Tested?
This has been tested with the [Gradle Enterprise build validation scripts](https://github.com/gradle/gradle-enterprise-build-validation-scripts/blob/main/Gradle.md).

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Delete as appropriate -->
- Non-breaking improvements to the build

## Checklist:
- [x] My code follows the [code style][style-guide] of this project.
- [ ] I have updated the documentation accordingly.

[style-guide]: https://wiki.apache.org/jmeter/CodeStyleGuidelines

Commit:
Use the most recent version for the autostyle plugin

This version contains fixes to exclude empty directories for input
files that previously caused issues with incremental builds because
some empty directories are created during a build. Subsequent builds
then had different input files and where no longer up to date.

Commit:
Exclude an empty directory from the zip and rat tasks

This empty directory that is created during builds would cause
subsequent builds to execute the tasks again because they would no
longer be up to date.

Commit:
Increase heap size

Total heap size should be bigger than max metaspace size which is part
of the heap.

Commit:
Ignore empty directories for batch test task input directory

Empty directories are created in bin/testfiles that need to be ignored
as task inputs to improve incremental builds.

Commit:
Add task dependencies for the rat task

This enables incremental build support for this task.

Commit:
Enabled build caching for some tasks

Commit:
Use relative paths in task inputs

This improves build caching because those paths are part of the cache
key calculation.

runningcode commented on src/licenses/build.gradle.kts:
```diff
@@ -73,6 +74,7 @@ val gatherBinaryLicenses by tasks.registering(GatherLicenseTask::class) {
     configuration(binaryDependencies)
     ignoreMissingLicenseFor.add(SpdxLicense.Apache_2_0.asExpression())
     defaultTextFor.add(SpdxLicense.MPL_2_0.asExpression())
+    outputs.cacheIf { true }
```
Same comment as above. Do you know more about this task? is it a disk bound task where it just copies a bunch of license files or is there some more processing involved here to merge them or something like that?

runningcode commented on src/dist/build.gradle.kts:
```diff
@@ -439,9 +440,10 @@ fun xslt(
+    inputs.files(xdocs).withPathSensitivity(PathSensitivity.RELATIVE).withPropertyName("xdocs")
     inputs.property("year", lastEditYear)
     outputs.dir(outputDir)
+    outputs.cacheIf { true }
```
Do you know more about this task? is it a disk bound task where it just copies a bunch of license files or is there some more processing involved here to merge them or something like that?

sandra-thieme replied:
It looks like it's doing more than just copying files (reading manifests and poms, ..), but I cannot confidently say that this change won't break anything. So maybe it's better to remove it.

runningcode replied:
Let's remove it then!

sandra-thieme replied:
Removed.

runningcode commented on build.gradle.kts:
```diff
@@ -329,9 +331,7 @@ allprojects {
-                configProperties = mapOf(
-                    "cache_file" to buildDir.resolve("checkstyle/cacheFile")
-                )
+                configDirectory.set(rootProject.file("config/checkstyle"))
```
I think this change was good for learning but let's not upstream this change right now.

sandra-thieme replied:
I removed this change

runningcode reviewed (commented):
Looks good. In general, when marking a task as cacheable, we need to be sure that it is a CPU bound task and will benefit from caching and we need to ensure that all the inputs and outputs are properly tracked. If just a single input or output is missing it can lead to incorrect builds.

runningcode reviewed (approved):
LGTM! when upstreaming the change please include a build scan link on scans.gradle.com illustrating something that was fixed. :)

Commit:
Use newest version of vlsi-release-plugins

1.85 contains a fix for the license gather plugin switching absolute to
relative paths to better support build caching.
